### ISSUE TITLE
"chomp @a = <STDIN>" not an error?

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5589]
Migrated from [rt.perl.org#9734](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=9734) (status was 'resolved')

Searchable as RT9734$

**Comment by p5pRT** [comment]
### From @RandalSchwartz


OK\, I'm puzzled\.

&nbsp;   chomp $a = \<STDIN>;

rightfully aborts as a syntax error\.  But

&nbsp;   chomp @&#8203;a = \<STDIN>;

doesn't\, and does something odd \(nothing is in @&#8203;a\, and \<STDIN> is
read in scalar context\)\.  Shouldn't this be a syntax error?

Even in nearly\-bleed\-perl\, it deparses as&#8203;:

&nbsp;    merlyn% perl \-MO=Deparse\,\-p \-e 'chomp @&#8203;a = \<STDIN>'
&nbsp;    \(chomp\(@&#8203;a\) = \<STDIN>\);
&nbsp;    \-e syntax OK
&nbsp;    merlyn%

which still looks like bad syntax to me\.

\(Found by one of my students\, who didn't know better\. :\)

\-\- 
Randal L\. Schwartz \- Stonehenge Consulting Services\, Inc\. \- \+1 503 777 0095
\<merlyn@&#8203;stonehenge\.com> \<URL&#8203;:http&#8203;://www\.stonehenge\.com/merlyn/>
Perl/Unix/security consulting\, Technical writing\, Comedy\, etc\. etc\.
See PerlTraining\.Stonehenge\.com for onsite and open\-enrollment Perl training\!

**Comment by p5pRT** [comment]
### From @lizmat

At 01&#8203;:38 PM 6/19/02 \-0700\, Randal L\. Schwartz wrote&#8203;:
>OK\, I'm puzzled\.
>
>    chomp $a = \<STDIN>;
>
>rightfully aborts as a syntax error\.  But
>
>    chomp @&#8203;a = \<STDIN>;
>
>doesn't\, and does something odd \(nothing is in @&#8203;a\, and \<STDIN> is
>read in scalar context\)\.  Shouldn't this be a syntax error?
>
>Even in nearly\-bleed\-perl\, it deparses as&#8203;:
>
>     merlyn% perl \-MO=Deparse\,\-p \-e 'chomp @&#8203;a = \<STDIN>'
>     \(chomp\(@&#8203;a\) = \<STDIN>\);
>     \-e syntax OK
>     merlyn%
>
>which still looks like bad syntax to me\.
>
>\(Found by one of my students\, who didn't know better\. :\)

According to perlfunc&#8203;:

=========================================================================
Any function in the list below may be used either with or without 
parentheses around its arguments\.
=========================================================================

and chomp \_does\_ occur in that list\.  And from the docco of chomp\(\)&#8203;:

=========================================================================
You can actually chomp anything that's an lvalue\, including an assignment&#8203;:
&nbsp;     chomp\($cwd = \`pwd\`\);
&nbsp;     chomp\($answer = \<STDIN>\);
If you chomp a list\, each element is chomped\, and the total number of 
characters removed is returned\.
=========================================================================

which would lead me to conclude that either both should be legal\, or the 
documentation should be changed\.



Liz

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

\-\-\-\-\- Original Message \-\-\-\-\-
From&#8203;: "Elizabeth Mattijsen" \<liz@&#8203;dijkmat\.nl>
To&#8203;: "Randal L\. Schwartz" \<merlyn@&#8203;stonehenge\.com>; \<perl5\-porters@&#8203;perl\.org>
Sent&#8203;: Sunday\, June 23\, 2002 5&#8203;:43 PM
Subject&#8203;: Re&#8203;: \[ID 20020623\.015\] "chomp @&#8203;a = \<STDIN>" not an error?


> At 01&#8203;:38 PM 6/19/02 \-0700\, Randal L\. Schwartz wrote&#8203;:
> =========================================================================
> You can actually chomp anything that's an lvalue\, including an assignment&#8203;:
>      chomp\($cwd = \`pwd\`\);
>      chomp\($answer = \<STDIN>\);
> If you chomp a list\, each element is chomped\, and the total number of
> characters removed is returned\.
> =========================================================================
>
> which would lead me to conclude that either both should be legal\, or the
> documentation should be changed\.

Yes\, but chomp binds tighter than assignment\, as the Deparse output
indicates\, so you're trying to assign to the non\-lvalue return of chomp\.
Thus the parentheses in the example\.

Benjamin Holzman

**Comment by p5pRT** [comment]
### From @RandalSchwartz

>>>>> "You" ==   \<bholzman@&#8203;earthlink\.net> writes&#8203;:

>> At 01&#8203;:38 PM 6/19/02 \-0700\, Randal L\. Schwartz wrote&#8203;:
>> =========================================================================
>> You can actually chomp anything that's an lvalue\, including an assignment&#8203;:
>> chomp\($cwd = \`pwd\`\);
>> chomp\($answer = \<STDIN>\);
>> If you chomp a list\, each element is chomped\, and the total number of
>> characters removed is returned\.
>> =========================================================================
>> 
>> which would lead me to conclude that either both should be legal\, or the
>> documentation should be changed\.

You> Yes\, but chomp binds tighter than assignment\, as the Deparse output
You> indicates\, so you're trying to assign to the non\-lvalue return of chomp\.
You> Thus the parentheses in the example\.

Right\, so

&nbsp;        chomp $foo = \<BAR>

is clearly illegal\, because it wants to be parsed as&#8203;:

&nbsp;        \(chomp $foo\) = \<BAR>

My ponderance is why

&nbsp;        chomp @&#8203;foo = \<BAR>

is not also parsed identically\, as the illegal&#8203;:

&nbsp;        \(chomp @&#8203;foo\) = \<BAR>

The relative precedence of chomp\(\) and "=" doesn't change\, so the
parsing is wrong\.  Or something\.  I just don't get it\.

\-\- 
Randal L\. Schwartz \- Stonehenge Consulting Services\, Inc\. \- \+1 503 777 0095
\<merlyn@&#8203;stonehenge\.com> \<URL&#8203;:http&#8203;://www\.stonehenge\.com/merlyn/>
Perl/Unix/security consulting\, Technical writing\, Comedy\, etc\. etc\.
See PerlTraining\.Stonehenge\.com for onsite and open\-enrollment Perl training\!

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]


\-\-\-\-\- Original Message \-\-\-\-\- 
From&#8203;: "Randal L\. Schwartz" \<merlyn@&#8203;stonehenge\.com>
To&#8203;: \<bholzman@&#8203;earthlink\.net>
Cc&#8203;: \<perl5\-porters@&#8203;perl\.org>; "Elizabeth Mattijsen" \<liz@&#8203;dijkmat\.nl>
Sent&#8203;: Monday\, June 24\, 2002 7&#8203;:48 PM
Subject&#8203;: Re&#8203;: \[ID 20020623\.015\] "chomp @&#8203;a = \<STDIN>" not an error?
> Right\, so
> 
>         chomp $foo = \<BAR>
> 
> is clearly illegal\, because it wants to be parsed as&#8203;:
> 
>         \(chomp $foo\) = \<BAR>
> 
> My ponderance is why
> 
>         chomp @&#8203;foo = \<BAR>
> 
> is not also parsed identically\, as the illegal&#8203;:
> 
>         \(chomp @&#8203;foo\) = \<BAR>
> 
> The relative precedence of chomp\(\) and "=" doesn't change\, so the
> parsing is wrong\.  Or something\.  I just don't get it\.

Oh yeah\, I totally agree that it should be a syntax error\.

**Comment by p5pRT** [comment]
### From @pjscott

At 04&#8203;:48 PM 6/24/02 \-0700\, Randal L\. Schwartz wrote&#8203;:
>Right\, so
>
>         chomp $foo = \<BAR>
>
>is clearly illegal\, because it wants to be parsed as&#8203;:
>
>         \(chomp $foo\) = \<BAR>
>
>My ponderance is why
>
>         chomp @&#8203;foo = \<BAR>
>
>is not also parsed identically\, as the illegal&#8203;:
>
>         \(chomp @&#8203;foo\) = \<BAR>
>
>The relative precedence of chomp\(\) and "=" doesn't change\, so the
>parsing is wrong\.  Or something\.  I just don't get it\.

Perl parses scalar chomps and list chomps as different operators \(the 
former is called schomp\)\.

% perl \-e 'chomp\($a\) = 42'
Can't modify scalar chomp in scalar assignment at \-e line 1\, at EOF
Execution of \-e aborted due to compilation errors\.
% perl \-e 'chomp\(@&#8203;a\) = 42'
%

I spent a fair bit of time digging around op\.c\, toke\.c and opcode\.pl 
trying to figure out what schomp does that chomp doesn't \(or vice 
versa\)\.  All I can tell is that during the OP\_SASSIGN\, there's no case 
statement in Perl\_mod\(\) in op\.c for OP\_SCHOMP\, so it ends up in the 
default block\, which triggers the error\.  But there is a case statement 
for OP\_CHOMP\, and that ends up getting called and has no objection\.

Maybe adding something like

@&#8203;@&#8203; \-1613\,6 \+1613\,8 @&#8203;@&#8203;
&nbsp;      case OP\_NEXTSTATE&#8203;:
&nbsp;      case OP\_DBSTATE&#8203;:
&nbsp;      case OP\_CHOMP&#8203;:
\+       if \(type == OP\_SASSIGN\)
\+        goto nomod;
&nbsp;         PL\_modcount = RETURN\_UNLIMITED\_NUMBER;
&nbsp;         break;
&nbsp;      case OP\_RV2SV&#8203;:

?  But this seems kludgey\.

\-\-
Peter Scott
Pacific Systems Design Technologies

**Comment by p5pRT** [comment]
### From @tamias

On Tue\, Jun 25\, 2002 at 11&#8203;:44&#8203;:29AM \+0200\, Elizabeth Mattijsen wrote&#8203;:
> \-\-\- perlfunc\.pod\.original    Fri Jun 21 19&#8203;:38&#8203;:36 2002
> \+\+\+ perlfunc\.pod    Tue Jun 25 11&#8203;:43&#8203;:19 2002
> @&#8203;@&#8203; \-621\,7 \+621\,7 @&#8203;@&#8203;
>  
>  =item chomp VARIABLE
>  
> \-=item chomp\( LIST \)
> \+=item chomp\( LIST \)   \# note parentheses here
>  
>  =item chomp
>  
> @&#8203;@&#8203; \-644\,30 \+644\,39 @&#8203;@&#8203;
>      \}
>  
>  If VARIABLE is a hash\, it chomps the hash's values\, but not its keys\.
> \+If VARIABLE is an array\, each element is chomped\.  In both cases the total
> \+number of characters removed is returned\.
> 
>  You can actually chomp anything that's an lvalue\, including an assignment&#8203;:
>  
>      chomp\($cwd = \`pwd\`\);
>      chomp\($answer = \<STDIN>\);
>  
> \-If you chomp a list\, each element is chomped\, and the total number of
> \-characters removed is returned\.

Should we still explicitly document the behavior when chomping a list\, as
distinct from chomping an array?

> \+Note that parentheses are necessary when you're chomping anything that is
> \+not a simple variable\.  This is because the chomp\(\) has a higher precedence
> \+than the assignment\, so that the "chomp $cwd = \`pwd\`;" is interpreted as
> \+"\(chomp $cwd\) = \`pwd\`;"\, rather than as "chomp \($cwd = \`pwd\`\)" which you
> \+might expect\. 

The second sentence refers to "the assignment"\, which is a bit odd after
the new first sentence\, which no longer specifically mentions an
assignment\.  How about&#8203;:

&nbsp;  Note that parentheses are necessary when you're chomping anything that is
&nbsp;  not a simple variable\.  This is because chomp\(\) has the highest precedence\.
&nbsp;  For example\, C\<chomp $cwd = \`pwd\`> is interpreted as C\<\(chomp $cwd\) =
&nbsp;  \`pwd\`> rather than C\<chomp\($cwd = \`pwd\`> as you might expect\.

I don't know if "the highest precedence" is the right way to say it
though\.\.\.


>  =item chop VARIABLE
>  
> \-=item chop\( LIST \)
> \+=item chop\( LIST \)     \# note parentheses here
>  
>  =item chop
>  
>  Chops off the last character of a string and returns the character
>  chopped\.  It is much more efficient than C\<s/\.$//s> because it neither
>  scans nor copies the string\.  If VARIABLE is omitted\, chops C\<$\_>\.
> \+
>  If VARIABLE is a hash\, it chops the hash's values\, but not its keys\.
> \+If VARIABLE is an array\, each element is chopped\.  In all cases the value of
> \+the last C\<chop> is returned\.
>  
>  You can actually chop anything that's an lvalue\, including an assignment\.
> \-
> \-If you chop a list\, each element is chopped\.  Only the value of the
> \-last C\<chop> is returned\.
> \+But note that parentheses are necessary when you're chopping anything that
> \+is not a simple variable\.  This is because the chop\(\) has a higher precedence
> \+than the assignment\, so that the "chop $a\,$b;" is interpreted as
> \+"\(chop $a\)\,$b;" rather than as "chop \($a\,$b\)" which you might expect\. 
>  
>  Note that C\<chop> returns the last character\.  To return all but the last
>  character\, use C\<substr\($string\, 0\, \-1\)>\.

Similar comments for this part as well\.  :\)

Ronald

**Comment by p5pRT** [comment]
### From @lizmat

At 04&#8203;:48 PM 6/24/02 \-0700\, Randal L\. Schwartz wrote&#8203;:
> >> which would lead me to conclude that either both should be legal\, or the
> >> documentation should be changed\.
>You> Yes\, but chomp binds tighter than assignment\, as the Deparse output
>You> indicates\, so you're trying to assign to the non\-lvalue return of chomp\.
>You> Thus the parentheses in the example\.

In any case\, the documentation of chop\(\) and chomp\(\) may need 
clarification\.  Document patch is included\.


Liz

**Comment by p5pRT** [comment]
### From @lizmat

<details><summary>perlfunc.patch</summary>

```diff
--- perlfunc.pod.original	Fri Jun 21 19:38:36 2002
+++ perlfunc.pod	Tue Jun 25 10:26:41 2002
@@ -650,6 +650,11 @@
     chomp($cwd = `pwd`);
     chomp($answer = <STDIN>);
 
+Note that parentheses are necessary when you're chomping a left-value.  This
+is because the chomp() has a higher precedence than the assignment, so that
+the "chomp $cwd = `pwd`;" is interpreted as "(chomp $cwd) = `pwd`;", rather
+than as "chomp ($cwd = `pwd`)" which you might expect. 
+
 If you chomp a list, each element is chomped, and the total number of
 characters removed is returned.
 
@@ -665,6 +670,10 @@
 If VARIABLE is a hash, it chops the hash's values, but not its keys.
 
 You can actually chop anything that's an lvalue, including an assignment.
+But note that parentheses are necessary when you're chopping a left-value.
+This is because the chop() has a higher precedence than the assignment, so that
+the "chop $line = <STDIN>;" is interpreted as "(chop $line) = <STDIN>;" rather
+than as "chop ($line = <STDIN>)" which you might expect. 
 
 If you chop a list, each element is chopped.  Only the value of the
 last C<chop> is returned.

```
</details>

**Comment by p5pRT** [comment]
### From @lizmat

At 10&#8203;:59 AM 6/25/02 \-0400\, Ronald J Kimball wrote&#8203;:
>On Tue\, Jun 25\, 2002 at 11&#8203;:44&#8203;:29AM \+0200\, Elizabeth Mattijsen wrote&#8203;:
> > \-If you chomp a list\, each element is chomped\, and the total number of
> > \-characters removed is returned\.
>Should we still explicitly document the behavior when chomping a list\, as
>distinct from chomping an array?

In my book an array is a variable such as @&#8203;list\, something that can be used 
as a left value\.  An array can be interpreted as a list in list context\.  A 
list is any range of values\, such as \(1\,2\,3\)\, qw\(a b c\) or 1\.\.20\.

Are these definitions wrong?


> > \+Note that parentheses are necessary when you're chomping anything that is
> > \+not a simple variable\.  This is because the chomp\(\) has a higher 
> precedence
> > \+than the assignment\, so that the "chomp $cwd = \`pwd\`;" is interpreted as
> > \+"\(chomp $cwd\) = \`pwd\`;"\, rather than as "chomp \($cwd = \`pwd\`\)" which you
> > \+might expect\.
>
>The second sentence refers to "the assignment"\, which is a bit odd after
>the new first sentence\, which no longer specifically mentions an
>assignment\.  How about&#8203;:
>
>   Note that parentheses are necessary when you're chomping anything that is
>   not a simple variable\.  This is because chomp\(\) has the highest precedence\.
>   For example\, C\<chomp $cwd = \`pwd\`> is interpreted as C\<\(chomp $cwd\) =
>   \`pwd\`> rather than C\<chomp\($cwd = \`pwd\`> as you might expect\.
>
>I don't know if "the highest precedence" is the right way to say it
>though\.\.\.

Let's not confuse users in that respect\.  Third version of doc patch included\.



Liz

**Comment by p5pRT** [comment]
### From @lizmat

<details><summary>perlfunc3.patch</summary>

```diff
--- perlfunc.pod.original	Fri Jun 21 19:38:36 2002
+++ perlfunc.pod	Tue Jun 25 17:17:46 2002
@@ -621,7 +621,7 @@
 
 =item chomp VARIABLE
 
-=item chomp( LIST )
+=item chomp( LIST )   # note parentheses here
 
 =item chomp
 
@@ -644,30 +644,37 @@
     }
 
 If VARIABLE is a hash, it chomps the hash's values, but not its keys.
+If VARIABLE is an array, each element is chomped.  In both cases the total
+number of characters removed is returned.
 
 You can actually chomp anything that's an lvalue, including an assignment:
 
     chomp($cwd = `pwd`);
     chomp($answer = <STDIN>);
 
-If you chomp a list, each element is chomped, and the total number of
-characters removed is returned.
+Note that parentheses are necessary when you're chomping anything that is
+not a simple variable.  This is because "chomp $cwd = `pwd`;" is interpreted as
+"(chomp $cwd) = `pwd`;", rather than as "chomp( $cwd = `pwd` )" which you
+might expect. 
 
 =item chop VARIABLE
 
-=item chop( LIST )
+=item chop( LIST )     # note parentheses here
 
 =item chop
 
 Chops off the last character of a string and returns the character
 chopped.  It is much more efficient than C<s/.$//s> because it neither
 scans nor copies the string.  If VARIABLE is omitted, chops C<$_>.
+
 If VARIABLE is a hash, it chops the hash's values, but not its keys.
+If VARIABLE is an array, each element is chopped.  In all cases the value of
+the last C<chop> is returned.
 
 You can actually chop anything that's an lvalue, including an assignment.
-
-If you chop a list, each element is chopped.  Only the value of the
-last C<chop> is returned.
+But note that parentheses are necessary when you're chopping anything that
+is not a simple variable.  This is because "chop $a,$b;" is interpreted as
+"(chop $a),$b;" rather than as "chop( $a,$b )" which you might expect. 
 
 Note that C<chop> returns the last character.  To return all but the last
 character, use C<substr($string, 0, -1)>.

```
</details>

**Comment by p5pRT** [comment]
### From @lizmat

At 11&#8203;:38 AM 6/25/02 \-0400\, Ronald J Kimball wrote&#8203;:
>For code snippets in POD\, C\<> is better than ""\.  That way you'll get the
>appropriate formatting in whatever renderer \(text\, html\, \*roff\, etc\.\) is
>being used\.

I've looked at the pod2html output\, but I find the C\<> syntax handler not 
necessarily clearer than using quotes\.  I guess I could create seperate 
lines for them\, making it \_really\_ clear that it is a code example\.

If the C\<> syntax is customary\, I'll supply yet another patch with those 
changes\.\.\.


Liz

**Comment by p5pRT** [comment]
### From @tamias

On Tue\, Jun 25\, 2002 at 05&#8203;:19&#8203;:33PM \+0200\, Elizabeth Mattijsen wrote&#8203;:
> At 10&#8203;:59 AM 6/25/02 \-0400\, Ronald J Kimball wrote&#8203;:
> >On Tue\, Jun 25\, 2002 at 11&#8203;:44&#8203;:29AM \+0200\, Elizabeth Mattijsen wrote&#8203;:
> >> \-If you chomp a list\, each element is chomped\, and the total number of
> >> \-characters removed is returned\.
> >Should we still explicitly document the behavior when chomping a list\, as
> >distinct from chomping an array?
> 
> In my book an array is a variable such as @&#8203;list\, something that can be used 
> as a left value\.  An array can be interpreted as a list in list context\.  A 
> list is any range of values\, such as \(1\,2\,3\)\, qw\(a b c\) or 1\.\.20\.
> 
> Are these definitions wrong?

An array can be a list\, but a list isn't necessarily an array\.  I'm just
wondering if the documentation should still make it explicit what happens
when you chomp a list that isn't an array\, such as chomp\($x\, $y\)\.

> >The second sentence refers to "the assignment"\, which is a bit odd after
> >the new first sentence\, which no longer specifically mentions an
> >assignment\.  How about&#8203;:
> >
> >  Note that parentheses are necessary when you're chomping anything that is
> >  not a simple variable\.  This is because chomp\(\) has the highest 
> >  precedence\.
> >  For example\, C\<chomp $cwd = \`pwd\`> is interpreted as C\<\(chomp $cwd\) =
> >  \`pwd\`> rather than C\<chomp\($cwd = \`pwd\`> as you might expect\.
> >
> >I don't know if "the highest precedence" is the right way to say it
> >though\.\.\.
> 
> Let's not confuse users in that respect\.  Third version of doc patch 
> included\.

Fair enough\.

> \+Note that parentheses are necessary when you're chomping anything that is
> \+not a simple variable\.  This is because "chomp $cwd = \`pwd\`;" is interpreted as
> \+"\(chomp $cwd\) = \`pwd\`;"\, rather than as "chomp\( $cwd = \`pwd\` \)" which you
> \+might expect\. 

For code snippets in POD\, C\<> is better than ""\.  That way you'll get the
appropriate formatting in whatever renderer \(text\, html\, \*roff\, etc\.\) is
being used\.

Ronald

**Comment by p5pRT** [comment]
### From @rgarcia

On 2002\.06\.25 10&#8203;:28 Elizabeth Mattijsen wrote&#8203;:
> In any case\, the documentation of chop\(\) and chomp\(\) may need 
> clarification\.  Document patch is included\.
> 
> 
> Liz
> \-\-\- perlfunc\.pod\.original    Fri Jun 21 19&#8203;:38&#8203;:36 2002
> \+\+\+ perlfunc\.pod    Tue Jun 25 10&#8203;:26&#8203;:41 2002
> @&#8203;@&#8203; \-650\,6 \+650\,11 @&#8203;@&#8203;
>      chomp\($cwd = \`pwd\`\);
>      chomp\($answer = \<STDIN>\);
>  
> \+Note that parentheses are necessary when you're chomping a left\-value\.  This
> \+is because the chomp\(\) has a higher precedence than the assignment\, so that
> \+the "chomp $cwd = \`pwd\`;" is interpreted as "\(chomp $cwd\) = \`pwd\`;"\, rather
> \+than as "chomp \($cwd = \`pwd\`\)" which you might expect\. 

In fact\, parentheses are always needed\, because the parsing of chom?p is
peculiar\. For example\, "chomp $x\, $y" really means "chomp\($x\)\,$y"\. It
doesn't behave as a real list operator \(unlike kill\(\)\, for example\,
whose equivalent perl prototype is "\(@&#8203;\)"\)\.

That's what the header in perlfunc tries to say :

=item chomp VARIABLE

=item chomp\( LIST \)

=item chomp

**Comment by p5pRT** [comment]
### From @lizmat

At 11&#8203;:26 AM 6/25/02 \+0200\, Rafael Garcia\-Suarez wrote&#8203;:
>On 2002\.06\.25 10&#8203;:28 Elizabeth Mattijsen wrote&#8203;:
> > In any case\, the documentation of chop\(\) and chomp\(\) may need
> > clarification\.  Document patch is included\.
>In fact\, parentheses are always needed\, because the parsing of chom?p is
>peculiar\. For example\, "chomp $x\, $y" really means "chomp\($x\)\,$y"\. It
>doesn't behave as a real list operator \(unlike kill\(\)\, for example\,
>whose equivalent perl prototype is "\(@&#8203;\)"\)\.
>
>That's what the header in perlfunc tries to say :
>
>=item chomp VARIABLE
>
>=item chomp\( LIST \)
>
>=item chomp

Because parentheses are sometimes used in documentation and sometimes not\, 
that pointer was \_completely\_ missed by me\.  Second version of document 
patch included\.  Hopefully the emphasis will now not be missed by users\.\.\.  ;\-\)


Liz

**Comment by p5pRT** [comment]
### From @lizmat

<details><summary>perlfunc2.patch</summary>

```diff
--- perlfunc.pod.original	Fri Jun 21 19:38:36 2002
+++ perlfunc.pod	Tue Jun 25 11:43:19 2002
@@ -621,7 +621,7 @@
 
 =item chomp VARIABLE
 
-=item chomp( LIST )
+=item chomp( LIST )   # note parentheses here
 
 =item chomp
 
@@ -644,30 +644,39 @@
     }
 
 If VARIABLE is a hash, it chomps the hash's values, but not its keys.
+If VARIABLE is an array, each element is chomped.  In both cases the total
+number of characters removed is returned.
 
 You can actually chomp anything that's an lvalue, including an assignment:
 
     chomp($cwd = `pwd`);
     chomp($answer = <STDIN>);
 
-If you chomp a list, each element is chomped, and the total number of
-characters removed is returned.
+Note that parentheses are necessary when you're chomping anything that is
+not a simple variable.  This is because the chomp() has a higher precedence
+than the assignment, so that the "chomp $cwd = `pwd`;" is interpreted as
+"(chomp $cwd) = `pwd`;", rather than as "chomp ($cwd = `pwd`)" which you
+might expect. 
 
 =item chop VARIABLE
 
-=item chop( LIST )
+=item chop( LIST )     # note parentheses here
 
 =item chop
 
 Chops off the last character of a string and returns the character
 chopped.  It is much more efficient than C<s/.$//s> because it neither
 scans nor copies the string.  If VARIABLE is omitted, chops C<$_>.
+
 If VARIABLE is a hash, it chops the hash's values, but not its keys.
+If VARIABLE is an array, each element is chopped.  In all cases the value of
+the last C<chop> is returned.
 
 You can actually chop anything that's an lvalue, including an assignment.
-
-If you chop a list, each element is chopped.  Only the value of the
-last C<chop> is returned.
+But note that parentheses are necessary when you're chopping anything that
+is not a simple variable.  This is because the chop() has a higher precedence
+than the assignment, so that the "chop $a,$b;" is interpreted as
+"(chop $a),$b;" rather than as "chop ($a,$b)" which you might expect. 
 
 Note that C<chop> returns the last character.  To return all but the last
 character, use C<substr($string, 0, -1)>.

```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

> \[RT\_System \- Wed Jun 26 00&#8203;:16&#8203;:06 2002\]&#8203;:
> 
> At 11&#8203;:26 AM 6/25/02 \+0200\, Rafael Garcia\-Suarez wrote&#8203;:
> >On 2002\.06\.25 10&#8203;:28 Elizabeth Mattijsen wrote&#8203;:
> > > In any case\, the documentation of chop\(\) and chomp\(\) may need
> > > clarification\.  Document patch is included\.
> >In fact\, parentheses are always needed\, because the parsing of chom?p
> is
> >peculiar\. For example\, "chomp $x\, $y" really means "chomp\($x\)\,$y"\. It
> >doesn't behave as a real list operator \(unlike kill\(\)\, for example\,
> >whose equivalent perl prototype is "\(@&#8203;\)"\)\.
> >
> >That's what the header in perlfunc tries to say :
> >
> >=item chomp VARIABLE
> >
> >=item chomp\( LIST \)
> >
> >=item chomp
> 
> Because parentheses are sometimes used in documentation and sometimes
> not\,
> that pointer was \_completely\_ missed by me\.  Second version of
> document
> patch included\.  Hopefully the emphasis will now not be missed by
> users\.\.\.  ;\-\)
> 
> 
> Liz
> 
> 

This documentation fix went into Perl as change \#17367\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
